internal/controllers/import_controller_v3.go

```diff
@@ -0,0 +1,365 @@
+/*
+Copyright © 2023 - 2024 SUSE LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	errorutils "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	"sigs.k8s.io/cluster-api/util/predicates"
+
+	managementv3 "github.com/rancher-sandbox/rancher-turtles/internal/rancher/management/v3"
+	"github.com/rancher-sandbox/rancher-turtles/util"
+	turtlespredicates "github.com/rancher-sandbox/rancher-turtles/util/predicates"
+)
+
+// CAPIImportManagementV3Reconciler represents a reconciler for importing CAPI clusters in Rancher.
+type CAPIImportManagementV3Reconciler struct {
+	Client             client.Client
+	RancherClient      client.Client
+	recorder           record.EventRecorder
+	WatchFilterValue   string
+	Scheme             *runtime.Scheme
+	InsecureSkipVerify bool
+
+	controller         controller.Controller
+	externalTracker    external.ObjectTracker
+	remoteClientGetter remote.ClusterClientGetter
```

Comment:
This structure is the same as previous CAPIImportReconciler. Why we need two?

Reply:
I don't think there is any harm in this instance.

Reply:
Using specific structs just for the sake of simplicity and readability.

Reply:
I agree, it’s just making it harder to review. There are many changes, but functionally there should be only: 
1. Apply v3 cluster instead of v1 
2. Watch on v3 cluster, map events to CAPI cluster via newly added labels on the resource
3. Collect registration token from the v3 cluster name (named) namespace.
4. Manually issue delete on both objects as their lifecycle is tied together in removal.
5. Optionally: Featuregate this functionality.
It is making difficult to verify there is only those change with so many code changes. Also because of that, there seems to be some test case duplication.